pkg/apis/resource/types.go

```diff
@@ -37,64 +286,326 @@ type ResourceClaim struct {
+	// could satisfy requests in this claim. It is ignored while allocating
+	// the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []ClaimConfiguration
+
+	// ControllerName defines the name of the DRA driver that is meant
+	// to handle allocation of this claim. If empty, allocation is handled
+	// by the scheduler while scheduling a pod.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
+	//
+	// +optional
+	ControllerName *string
+
+	// Future extension, ignored by older schedulers. This is fine because
+	// scoring allows users to define a preference, without making it a
+	// hard requirement.
+	//
+	// Score *SomeScoringStruct
 }
 
-// ResourceClaimStatus tracks whether the resource has been allocated and what
-// the resulting attributes are.
-type ResourceClaimStatus struct {
-	// DriverName is a copy of the driver name from the ResourceClass at
-	// the time when allocation started.
+// Request is a request for one of many resources required for a claim.
+// This is typically a request for a single resource like a device, but can
+// also ask for several identical devices. It might get extended to support
+// asking for one of several different alternatives.
+type Request struct {
+	// The name can be used to reference this request in a pod.spec.containers[].resources.claims
+	// entry and in a constraint of the claim.
+	//
+	// Must be a DNS label.
+	Name string
+
+	*RequestDetail
+
+	// FUTURE EXTENSION:
+	//
+	// OneOf contains a list of requests, only one of which must be satisfied.
+	// Requests are listed in order of priority.
+	//
 	// +optional
+	// +listType=atomic
+	// OneOf []RequestDetail
+}
+
+// RequestDetail is embedded inside Request. Exactly one field must be set.
+type RequestDetail struct {
+	// Device requests one or more devices.
+	//
+	// +required
+	Device *DeviceRequest
+}
+
+// DeviceRequest is currently the only permitted alternative in RequestDetail.
+type DeviceRequest struct {
+	// By referencing a DeviceClass, a request inherits additional
+	// configuration parameters and selectors.
+	//
+	// A class is required. Which classes are available depends on the cluster.
+	//
+	// Administrators may use this to restrict which devices may get
+	// requested by only installing classes with selectors for permitted
+	// devices. If users are free to request anything without restrictions,
+	// then an empty class called "none" can get created to permit
+	// `deviceClassName: none`.
+	//
+	// +required
+	DeviceClassName string
+
+	// Each selector must be satisfied by a device which is requested.
+	//
+	// +optional
+	// +listType=atomic
+	Selectors []Selector
+
+	// Amount determines how many devices to allocate for the request.
+	// The default if unset is exactly one device.
+	//
+	// +optional
+	Amount *Amount
```

Comment:
The downsides of `Amount *Amount` are:
- extra nesting (`amount: exactCount: 1`)
- implicit defaults and two ways for the same thing (`nil` = ``amount: exactCount: 1`)
- ugly "all" support (`amount: all: true`)

The alternative suggested in https://github.com/kubernetes/enhancements/pull/4709#discussion_r1640552271 is a `countMode` discriminator.